perfkitbenchmarker/configs/option_decoders.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2015 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes for verifying and decoding config option values."""
+
+import abc
+
+from perfkitbenchmarker import errors
+
+
+class ConfigOptionDecoder(object):
+  """Verifies and decodes a config option value.
+
+  Attributes:
+    component: string. Description of the component to which the option applies.
+    option: string. Name of the config option.
+    required: boolean. True if the config option is required. False if not.
+  """
+
+  def __init__(self, component, option, **kwargs):
+    """Initializes a ConfigOptionDecoder.
+
+    Args:
+      component: string. Description of the component to which the option
+          applies.
+      option: string. Name of the config option.
+      **kwargs: May optionally contain a 'default' key mapping to a value or
+          callable object. If a value is provided, the config option is
+          optional, and the provided value is the default if the user does not
+          set a value for the config option. If a callable object is provided,
+          the config option is optional, and the provided object is called to
+          determine the value if the user does not set a value for the config
+          option. If not provided, the config option is required.
+    """
+    self.component = component
+    self.option = option
+    self.required = 'default' not in kwargs
+    if not self.required:
+      self._default = kwargs.pop('default')
+    assert not kwargs, ('__init__() received unexpected keyword arguments: '
+                        '{0}'.format(kwargs))
+
+  @property
+  def default(self):
+    """Gets the config option's default value.
+
+    Returns:
+      Default value of an optional config option.
+    """
+    assert not self.required, (
+        'Attempted to get the default value of {0} required config option '
+        '"{1}".'.format(self.component, self.option))
+    if hasattr(self._default, '__call__'):
+      return self._default()
+    return self._default
+
+  @abc.abstractmethod
```

Comment:
As far as I can see, ConfigOptionDecoder's metaclass isn't abc.ABCMeta, but I'm pretty sure it needs to be if you want this to work properly. See https://docs.python.org/2/library/abc.html#abc.abstractmethod